homeassistant/components/konnected.py

```diff
@@ -0,0 +1,298 @@
+
+        _LOGGER.info('Storing data in hass.data[konnected]: %s', device_data)
+        self.hass.data[DOMAIN][CONF_DEVICES][self.device_id] = device_data
+
+    @property
+    def stored_configuration(self):
+        """Return the configuration stored in `hass.data` for this device."""
+        return self.hass.data[DOMAIN][CONF_DEVICES][self.device_id]
+
+    def sensor_configuration(self):
+        """Return the configuration map for syncing sensors."""
+        return [{'pin': p} for p in
+                self.stored_configuration[CONF_SENSORS].keys()]
+
+    def actuator_configuration(self):
+        """Return the configuration map for syncing actuators."""
+        return [{'pin': p,
+                 'trigger': (0 if data.get('activation') in [0, 'low'] else 1)}
+                for p, data in
+                self.stored_configuration[CONF_SWITCHES].items()]
+
+    def sync_device(self):
+        """Sync the new pin configuration to the Konnected device."""
+        desired_sensor_configuration = self.sensor_configuration()
+        current_sensor_configuration = [
+            {'pin': s[CONF_PIN]} for s in self.status.get('sensors')]
+        _LOGGER.info('%s: desired sensor config: %s', self.device_id,
+                     desired_sensor_configuration)
+        _LOGGER.info('%s: current sensor config: %s', self.device_id,
+                     current_sensor_configuration)
+
+        desired_actuator_config = self.actuator_configuration()
+        current_actuator_config = self.status.get('actuators')
+        _LOGGER.info('%s: desired actuator config: %s', self.device_id,
+                     desired_actuator_config)
+        _LOGGER.info('%s: current actuator config: %s', self.device_id,
+                     current_actuator_config)
+
+        if (desired_sensor_configuration != current_sensor_configuration) or \
+                (current_actuator_config != desired_actuator_config):
+            _LOGGER.info('pushing settings to device %s', self.device_id)
+            self.client.put_settings(
+                desired_sensor_configuration,
+                desired_actuator_config,
+                self.hass.data[DOMAIN].get('auth_token'),
+                self.hass.config.api.base_url + ENDPOINT_ROOT
+            )
+
+
+class KonnectedView(HomeAssistantView):
+    """View creates an endpoint to receive push updates from the device."""
+
+    url = UPDATE_ENDPOINT
+    name = 'api:konnected'
+    requires_auth = False  # Uses access token from configuration
+
+    def __init__(self, auth_token):
+        """Initialize the view."""
+        self.auth_token = auth_token
+
+    @asyncio.coroutine
+    def put(self, request: Request, device_id, pin_num, state) -> Response:
+        """Receive a sensor update via PUT request and async set state."""
+        hass = request.app['hass']
+        data = hass.data[DOMAIN]
+
+        auth = request.headers.get(AUTHORIZATION, None)
+        if 'Bearer {}'.format(self.auth_token) != auth:
```

Comment:
This is not safe to timing attacks. See [`hmac.compare_digest`](https://github.com/home-assistant/home-assistant/blob/321eb2ec6f077191e20a7fc07490a18df6f11f3f/homeassistant/components/http/auth.py#L34) for safe auth token checks.